File: retaRust/python_arch_reference/reta_architecture/completion_word.py
from __future__ import annotations

"""Stage-40 activated word-completion morphisms for Reta's prompt facade.

Historically ``libs/word_completerAlx.py`` owned a patched copy of a
``prompt_toolkit`` word completer.  Stage 40 moves that concrete matching logic
into the architecture package while the legacy module keeps the historical
``WordCompleter`` name as a compatibility facade.

Mathematical reading:

* the list/callable of completion words is a local completion section;
* the document prefix before the cursor is an input-open-set restriction;
* prefix/middle matching is a morphism from raw prompt text to completion
  candidates;
* yielded ``Completion`` objects form the local output section for the prompt;
* the old ``word_completerAlx.WordCompleter`` route and the new architecture
  route are connected by a natural transformation: both routes must yield the
  same completions for the same document context.
"""

from dataclasses import dataclass
from typing import Callable, Iterable, List, Mapping, Optional, Pattern, Sequence, Union

try:
    from prompt_toolkit.completion import CompleteEvent, Completer, Completion
    from prompt_toolkit.document import Document
    from prompt_toolkit.formatted_text import AnyFormattedText
except (ModuleNotFoundError, ImportError):
    class CompleteEvent:  # minimal fallback for -S architecture probes
        pass

    class Completer:
        def get_completions(self, document, complete_event):
            return ()

    class Completion:
        def __init__(self, text, start_position=0, display=None, display_meta=""):
            self.text = text
            self.start_position = start_position
            self.display = display if display is not None else text
            self.display_meta = display_meta

        def __repr__(self):
            return f"Completion({self.text!r}, start_position={self.start_position!r})"

    class Document:
        def __init__(self, text="", cursor_position=None):
            self.text = text
            self.cursor_position = len(text) if cursor_position is None else cursor_position

        @property
        def text_before_cursor(self):
            return self.text[: self.cursor_position]

        def get_word_before_cursor(self, WORD=False, pattern=None):
            before = self.text_before_cursor
            if pattern is not None:
                matches = list(pattern.finditer(before))
                return matches[-1].group(0) if matches else ""
            if WORD:
                parts = before.split()
                return parts[-1] if parts else ""
            index = len(before)
            while index > 0 and (before[index - 1].isalnum() or before[index - 1] == "_"):
                index -= 1
            return before[index:]

    AnyFormattedText = object


WordsSource = Union[List[str], Callable[[], List[str]]]


def resolve_words(words: WordsSource) -> List[str]:
    """Resolve a static or callable completion-word section."""
    if callable(words):
        return list(words())
    return list(words)


def word_before_cursor(document: Document, *, WORD: bool = False, sentence: bool = False, pattern: Optional[Pattern[str]] = None) -> str:
    """Return the prefix section inspected by the word-completion morphism."""
    if sentence:
        return document.text_before_cursor
    return document.get_word_before_cursor(WORD=WORD, pattern=pattern)


def word_completion_matches(word: str, prefix: str, *, ignore_case: bool = False, match_middle: bool = False) -> bool:
    """Return whether one completion word matches the current prefix.

    The implementation intentionally preserves the historical
    ``libs/word_completerAlx.py`` semantics, including the slightly unusual
    middle-match slice ``prefix[:len(word)] in word``.
    """
    candidate = word.lower() if ignore_case else word
    probe = prefix.lower() if ignore_case else prefix
    if match_middle:
        return probe[: len(candidate)] in candidate
    return candidate.startswith(probe[: len(candidate)])


def iter_word_completions(
    words: WordsSource,
    document: Document,
    *,
    ignore_case: bool = False,
    display_dict: Optional[Mapping[str, AnyFormattedText]] = None,
    meta_dict: Optional[Mapping[str, AnyFormattedText]] = None,
    WORD: bool = False,
    sentence: bool = False,
    match_middle: bool = False,
    pattern: Optional[Pattern[str]] = None,
) -> Iterable[Completion]:
    """Yield prompt-toolkit completions like the legacy ``WordCompleter``."""
    display_dict = display_dict or {}
    meta_dict = meta_dict or {}
    prefix = word_before_cursor(document, WORD=WORD, sentence=sentence, pattern=pattern)
    for word in resolve_words(words):
        if word_completion_matches(word, prefix, ignore_case=ignore_case, match_middle=match_middle):
            yield Completion(
                word,
                -len(prefix),
                display=display_dict.get(word, word),
                display_meta=meta_dict.get(word, ""),
            )


class ArchitectureWordCompleter(Completer):
    """Architecture-owned implementation behind ``word_completerAlx.WordCompleter``."""

    def __init__(
        self,
        words: WordsSource,
        ignore_case: bool = False,
        display_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        meta_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        WORD: bool = False,
        sentence: bool = False,
        match_middle: bool = False,
        pattern: Optional[Pattern[str]] = None,
    ) -> None:
        assert not (WORD and sentence)
        self.words = words
        self.ignore_case = ignore_case
        self.display_dict = display_dict or {}
        self.meta_dict = meta_dict or {}
        self.WORD = WORD
        self.sentence = sentence
        self.match_middle = match_middle
        self.pattern = pattern

    def __eq__(self, obj) -> bool:
        return self.words == obj.words

    def get_completions(self, document: Document, complete_event: CompleteEvent) -> Iterable[Completion]:
        yield from iter_word_completions(
            self.words,
            document,
            ignore_case=self.ignore_case,
            display_dict=self.display_dict,
            meta_dict=self.meta_dict,
            WORD=self.WORD,
            sentence=self.sentence,
            match_middle=self.match_middle,
            pattern=self.pattern,
        )


@dataclass(frozen=True)
class WordCompletionMorphismBundle:
    """Activated architecture owner for legacy word-completion matching."""

    legacy_owner: str = "libs.word_completerAlx"
    activated_stage: int = 40

    def create_completer(
        self,
        words: WordsSource,
        ignore_case: bool = False,
        display_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        meta_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        WORD: bool = False,
        sentence: bool = False,
        match_middle: bool = False,
        pattern: Optional[Pattern[str]] = None,
    ) -> ArchitectureWordCompleter:
        return ArchitectureWordCompleter(
            words,
            ignore_case=ignore_case,
            display_dict=display_dict,
            meta_dict=meta_dict,
            WORD=WORD,
            sentence=sentence,
            match_middle=match_middle,
            pattern=pattern,
        )

    def resolve_words(self, words: WordsSource) -> List[str]:
        return resolve_words(words)

    def word_before_cursor(self, document: Document, *, WORD: bool = False, sentence: bool = False, pattern: Optional[Pattern[str]] = None) -> str:
        return word_before_cursor(document, WORD=WORD, sentence=sentence, pattern=pattern)

    def matches(self, word: str, prefix: str, *, ignore_case: bool = False, match_middle: bool = False) -> bool:
        return word_completion_matches(word, prefix, ignore_case=ignore_case, match_middle=match_middle)

    def completions(
        self,
        words: WordsSource,
        document: Document,
        *,
        ignore_case: bool = False,
        display_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        meta_dict: Optional[Mapping[str, AnyFormattedText]] = None,
        WORD: bool = False,
        sentence: bool = False,
        match_middle: bool = False,
        pattern: Optional[Pattern[str]] = None,
    ) -> Iterable[Completion]:
        return iter_word_completions(
            words,
            document,
            ignore_case=ignore_case,
            display_dict=display_dict,
            meta_dict=meta_dict,
            WORD=WORD,
            sentence=sentence,
            match_middle=match_middle,
            pattern=pattern,
        )

    def sample_completions(self, prefix: str = "re") -> list[str]:
        return [item.text for item in self.completions(["reta", "religion", "alpha"], Document(prefix))]

    def snapshot(self) -> dict:
        return {
            "class": "WordCompletionMorphismBundle",
            "stage": self.activated_stage,
            "legacy_owner": self.legacy_owner,
            "capsule": "InputPromptCapsule",
            "category": "ActivatedWordCompletionCategory",
            "functor": "WordCompletionActivationFunctor",
            "natural_transformation": "WordCompleterToArchitectureTransformation",
            "morphisms": [
                "resolve_words",
                "word_before_cursor",
                "word_completion_matches",
                "iter_word_completions",
                "create_completer",
            ],
            "compatibility_names": [
                "WordCompleter",
                "word_completerAlx.WordCompleter",
            ],
            "sample": {
                "prefix": "re",
                "texts": self.sample_completions("re"),
            },
        }


def bootstrap_word_completion_morphisms() -> WordCompletionMorphismBundle:
    """Return the Stage-40 activated word-completion bundle."""
    return WordCompletionMorphismBundle()

File: retaRust/crates/reta_architecture/src/completion_word.rs
//! Word-completion morphisms transcompiled from
//! `python_arch_reference/reta_architecture/completion_word.py`.
//!
//! This mirrors the legacy `word_completerAlx.WordCompleter` matching rules in
//! a prompt-toolkit-free Rust section.  It gives `rretaPrompt` a typed local
//! completion morphism before the complete nested completer is replaced.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct PromptDocument {
    pub text: String,
    pub cursor_position: usize,
}

impl PromptDocument {
    pub fn new(text: impl Into<String>) -> Self {
        let text = text.into();
        let cursor_position = text.chars().count();
        Self {
            text,
            cursor_position,
        }
    }

    pub fn with_cursor(text: impl Into<String>, cursor_position: usize) -> Self {
        Self {
            text: text.into(),
            cursor_position,
        }
    }

    pub fn text_before_cursor(&self) -> String {
        self.text.chars().take(self.cursor_position).collect()
    }

    pub fn get_word_before_cursor(&self, word: bool) -> String {
        let before = self.text_before_cursor();
        if word {
            return before
                .split_whitespace()
                .last()
                .unwrap_or_default()
                .to_string();
        }
        let mut reversed = Vec::new();
        for ch in before.chars().rev() {
            if ch.is_alphanumeric() || ch == '_' {
                reversed.push(ch);
            } else {
                break;
            }
        }
        reversed.into_iter().rev().collect()
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct CompletionCandidate {
    pub text: String,
    pub start_position: isize,
    pub display: String,
    pub display_meta: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct WordCompletionOptions {
    pub ignore_case: bool,
    pub word: bool,
    pub sentence: bool,
    pub match_middle: bool,
    pub display_dict: BTreeMap<String, String>,
    pub meta_dict: BTreeMap<String, String>,
}

impl Default for WordCompletionOptions {
    fn default() -> Self {
        Self {
            ignore_case: false,
            word: false,
            sentence: false,
            match_middle: false,
            display_dict: BTreeMap::new(),
            meta_dict: BTreeMap::new(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct WordCompletionMorphismBundle {
    pub legacy_owner: String,
    pub activated_stage: u32,
}

impl WordCompletionMorphismBundle {
    pub fn resolve_words(&self, words: &[String]) -> Vec<String> {
        resolve_words(words)
    }

    pub fn word_before_cursor(
        &self,
        document: &PromptDocument,
        options: &WordCompletionOptions,
    ) -> String {
        word_before_cursor(document, options.word, options.sentence)
    }

    pub fn matches(&self, word: &str, prefix: &str, options: &WordCompletionOptions) -> bool {
        word_completion_matches(word, prefix, options.ignore_case, options.match_middle)
    }

    pub fn completions(
        &self,
        words: &[String],
        document: &PromptDocument,
        options: &WordCompletionOptions,
    ) -> Vec<CompletionCandidate> {
        iter_word_completions(words, document, options)
    }

    pub fn sample_completions(&self, prefix: &str) -> Vec<String> {
        self.completions(
            &[
                "reta".to_string(),
                "religion".to_string(),
                "alpha".to_string(),
            ],
            &PromptDocument::new(prefix),
            &WordCompletionOptions::default(),
        )
        .into_iter()
        .map(|item| item.text)
        .collect()
    }

    pub fn snapshot(&self) -> WordCompletionSnapshot {
        WordCompletionSnapshot {
            class: "WordCompletionMorphismBundle".to_string(),
            stage: self.activated_stage,
            legacy_owner: self.legacy_owner.clone(),
            capsule: "InputPromptCapsule".to_string(),
            category: "ActivatedWordCompletionCategory".to_string(),
            functor: "WordCompletionActivationFunctor".to_string(),
            natural_transformation: "WordCompleterToArchitectureTransformation".to_string(),
            morphisms: vec![
                "resolve_words".to_string(),
                "word_before_cursor".to_string(),
                "word_completion_matches".to_string(),
                "iter_word_completions".to_string(),
                "create_completer".to_string(),
            ],
            compatibility_names: vec![
                "WordCompleter".to_string(),
                "word_completerAlx.WordCompleter".to_string(),
            ],
            sample_prefix: "re".to_string(),
            sample_texts: self.sample_completions("re"),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct WordCompletionSnapshot {
    pub class: String,
    pub stage: u32,
    pub legacy_owner: String,
    pub capsule: String,
    pub category: String,
    pub functor: String,
    pub natural_transformation: String,
    pub morphisms: Vec<String>,
    pub compatibility_names: Vec<String>,
    pub sample_prefix: String,
    pub sample_texts: Vec<String>,
}

pub fn bootstrap_word_completion_morphisms() -> WordCompletionMorphismBundle {
    WordCompletionMorphismBundle {
        legacy_owner: "libs.word_completerAlx".to_string(),
        activated_stage: 40,
    }
}

pub fn resolve_words(words: &[String]) -> Vec<String> {
    words.to_vec()
}

pub fn word_before_cursor(document: &PromptDocument, word: bool, sentence: bool) -> String {
    if sentence {
        document.text_before_cursor()
    } else {
        document.get_word_before_cursor(word)
    }
}

pub fn word_completion_matches(
    word: &str,
    prefix: &str,
    ignore_case: bool,
    match_middle: bool,
) -> bool {
    let candidate = if ignore_case {
        word.to_lowercase()
    } else {
        word.to_string()
    };
    let probe = if ignore_case {
        prefix.to_lowercase()
    } else {
        prefix.to_string()
    };
    let prefix_slice = probe
        .chars()
        .take(candidate.chars().count())
        .collect::<String>();
    if match_middle {
        candidate.contains(&prefix_slice)
    } else {
        candidate.starts_with(&prefix_slice)
    }
}

pub fn iter_word_completions(
    words: &[String],
    document: &PromptDocument,
    options: &WordCompletionOptions,
) -> Vec<CompletionCandidate> {
    let prefix = word_before_cursor(document, options.word, options.sentence);
    let mut out = Vec::new();
    for word in resolve_words(words) {
        if word_completion_matches(&word, &prefix, options.ignore_case, options.match_middle) {
            out.push(CompletionCandidate {
                text: word.clone(),
                start_position: -(prefix.chars().count() as isize),
                display: options
                    .display_dict
                    .get(&word)
                    .cloned()
                    .unwrap_or_else(|| word.clone()),
                display_meta: options.meta_dict.get(&word).cloned().unwrap_or_default(),
            });
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn word_before_cursor_matches_fallback_document() {
        let doc = PromptDocument::new("alpha reta");
        assert_eq!(word_before_cursor(&doc, false, false), "reta");
        assert_eq!(word_before_cursor(&doc, false, true), "alpha reta");
    }

    #[test]
    fn middle_match_preserves_python_slice_shape() {
        assert!(word_completion_matches("religion", "lig", false, true));
        assert!(word_completion_matches(
            "religion",
            "religionXYZ",
            false,
            false
        ));
        assert!(!word_completion_matches("alpha", "re", false, false));
    }

    #[test]
    fn completions_use_negative_prefix_start_position() {
        let bundle = bootstrap_word_completion_morphisms();
        let completions = bundle.completions(
            &["reta".to_string(), "alpha".to_string()],
            &PromptDocument::new("re"),
            &WordCompletionOptions::default(),
        );
        assert_eq!(completions[0].text, "reta");
        assert_eq!(completions[0].start_position, -2);
    }
}
